# Project 1: Data Cleaning
Submitted on October 11, 2025

NLP1000 S17
Group Members:
- Aragon, Enrique
- Clavano, Angelica (Jack)
- Lozada, Job
- Alcantara, Van Asher

<span style="background-color:yellow; color:black">SAMPLE TABLE DELETE LATER</span>
| Column 1      | Column 2      |
| ------------- | ------------- |
| Cell 1, Row 1 | Cell 2, Row 1 |
| Cell 1, Row 2 | Cell 1, Row 2 |

## 1. Data Selection

<description here>

The specifications of the project requires (x) words, so we have decided to include the following books from the Bible: Matthew, Mark, Luke, and John. These are the Gospels of the Bible.

| Target Language      | Link to Source      |
| ------------- | ------------- | 
| Spanish  | https://www.bible.com/versions/1076-jbs-biblia-del-jubileo | |
| Tagalog | https://www.bible.com/versions/177-tlab-ang-biblia | |
| English | https://www.bible.com/versions/3523-nrsvue-new-revised-standard-version-updated-edition-2021 | |
| Hiligaynon/Ilonggo | https://www.bible.com/versions/2190-mbbhil12-maayong-balita-nga-biblia-2012 |
| Bikol/Bikolano | https://www.bible.com/versions/890-mbbbik92-marahay-na-bareta-biblia | |
| Waray | https://www.bible.com/versions/2198-mbbsam-samarenyo-meaning-based-bible-1984 | |
| Ilocano | https://www.bible.com/versions/782-ripv-ti-baro-a-naimbag-a-damag-biblia | |
| Cebuano | https://www.bible.com/versions/562-rcpv-ang-bag-ong-maayong-balita-biblia | |
| Kapampangan | https://www.bible.com/versions/1141-pmpv-ing-mayap-a-balita-biblia | |
| Pangasinense | https://www.bible.com/versions/2194-mbbpan83-maung-a-balita-biblia | |
| Yakan | https://www.bible.com/versions/1388-yakv-yakan | |
| Ivatan | https://www.bible.com/versions/1315-vtsp-ivatan | |
| Tausug  | https://www.bible.com/versions/1319-tsg-kitab-injil | |
| Yami  | https://www.bible.com/versions/2364-snt-seysyo-no-tao | |
| Tuwali Ifugao | https://www.bible.com/versions/2123-ifkwb-nan-kalin-apu-dios | |
| Masbateño/Masbatenyo | https://www.bible.com/versions/1222-msb-masbatenyo | |
| TOTAL | 16 | |

### 1.1 Web Scraping

Using `webscraper.py`, we were able to extract all the Gospels in the languages listed in Section 1. First, the bible abbreviation and number is compiled, then for each of the chapters and their chapter ranges (mat, mrk, luk, jhn) -- which are all listed the same on Bible.com -- it will scrape for the text and compile them in the `/data` folder. This may take around 3-4 min.

In [34]:
# import os
from bs4 import BeautifulSoup # beautifulsoup4
from urllib.request import urlopen
from pathlib import Path
import re
import pandas as pd

# pip install xlsxwriter

In [35]:
languages = ["spanish", "tagalog", "english", "hiligaynon", "bikol", "waray", "ilocano", "cebuano", "kapampangan", "pangasinense", "yakan", "ivatan", "tausug", "yami", "tuwali_ifugao", "masbateno"]
bibleNumbers = ["1076", "177", "3523", "2190", "890", "2198", "782", "562", "1141", "2194", "1388", "1315", "1319", "2364", "2123", "1222"]
bibleAbbreviation = ["JBS", "TLAB", "NRSVUE", "MBBHIL12", "MBBBIK92", "MBBSAM", "RIPV", "RCPV", "PMPV", "MBBPAN83", "YAKV", "VTSP", "TSG", "SNT", "IFKWB", "MSB"]

bookCodes = ["mat", "mrk", "luk", "jhn"]
chapterRanges = {
    "mat": [1, 28],
    "mrk": [1, 16],
    "luk": [1, 24],
    "jhn": [1, 21]
}

In [36]:
'''
# webscraper v4; adjusted to use 'Path' instead of 'os'
# instead of "!python webscraper_v3.py"
output_folder = "data"
data_folder = Path("data")
data_folder.mkdir(parents=True, exist_ok=True) 

# clear content of the file if it exists so it doesn't make copies
for lang in languages:
    file_path = data_folder / f"{lang}.txt"
    file_path.write_text("", encoding="utf-8")

for lang, bibleNumber, abbreviation in zip(languages, bibleNumbers, bibleAbbreviation):
  # if in the previous code block, languages, bibleNumbers, and/or bibleAbbreviation is "SKIP", just skip it
  if bibleNumber == "SKIP" or abbreviation == "SKIP":
        continue
  
  for bookCode, (start, end) in chapterRanges.items():
     for chapter in range(start, end + 1):
      urlPartOne = "https://www.bible.com/bible/"
      chapterNumber = str(chapter)
      bibleName = f".{abbreviation}"

      url = f"{urlPartOne}{bibleNumber}/{bookCode.upper()}.{chapterNumber}{bibleName}"
      print(f"Scraping URL: {url}")

      try: 
        page = urlopen(url)
        html = page.read().decode("utf-8")
        soup = BeautifulSoup(html, "html.parser")
        
        #IMPORTANT: THIS ONLY WORKS ON BIBLE.COM, for other sites just use inspect element then select the div/class containing the text so that it extracts just that
        text = soup.find('div', {'class': 'ChapterContent_reader__Dt27r'})

        if text:
          # excludes the pop-up notes
          for note in text.select('span.ChapterContent_note__YlDW0'):
            note.decompose()
          
          # excludes chapter headings
          for heading in text.select('span.ChapterContent_heading__xBDcs'):
            heading.decompose()
        
          # write to lang-specific file
          file_path = Path(output_folder) / f"{lang}.txt" # file_name = os.path.join(output_folder, f"{lang}.txt")

          with file_path.open("a", encoding="utf-8") as f: # with open(file_name, "a", encoding="utf-8") as f:
              f.write(text.get_text() + "\n") 
        else:
            print(f"No content found for {lang}, {bookCode}, chapter {chapter}")
      except Exception as e:
                  print(f"Error scraping {url}: {e}")
'''

'\n# webscraper v4; adjusted to use \'Path\' instead of \'os\'\n# instead of "!python webscraper_v3.py"\noutput_folder = "data"\ndata_folder = Path("data")\ndata_folder.mkdir(parents=True, exist_ok=True) \n\n# clear content of the file if it exists so it doesn\'t make copies\nfor lang in languages:\n    file_path = data_folder / f"{lang}.txt"\n    file_path.write_text("", encoding="utf-8")\n\nfor lang, bibleNumber, abbreviation in zip(languages, bibleNumbers, bibleAbbreviation):\n  # if in the previous code block, languages, bibleNumbers, and/or bibleAbbreviation is "SKIP", just skip it\n  if bibleNumber == "SKIP" or abbreviation == "SKIP":\n        continue\n  \n  for bookCode, (start, end) in chapterRanges.items():\n     for chapter in range(start, end + 1):\n      urlPartOne = "https://www.bible.com/bible/"\n      chapterNumber = str(chapter)\n      bibleName = f".{abbreviation}"\n\n      url = f"{urlPartOne}{bibleNumber}/{bookCode.upper()}.{chapterNumber}{bibleName}"\n      print(f

The text data as is will show up in the following format: 

```txt
Mateo 1

1An mga Ginikanan ni Jesu-Cristo(Lk. 3:23-38) 1Iyo ini an mga ginikanan ni Jesu-Cristo, na gikan ki...

Mateo 2

2An Pagdalaw kan mga Mago 1Namundag si Jesus sa Betlehem nin Judea, kan si Herodes an hade...
```

So...

```txt
(book)(space)(book number)
(new line)
(chapter)(title)(subtitle if any)(space)(line number)(content)
```

where the `(line number)(content)` repeats for the entire page, and then a new line is made for the next chapter to begin. This can be cleaned using RegEx.


## 2. Data Cleaning & Segmentation
Data Cleaning: *Remove document tags, unnecessary spaces, duplicated or repetitive terms, and unrelated symbols or keywords from the data.*

Segmentation: *Properly segment each paragraph into (a) verses and (b) sentences. An example is shown below.*

<span style="background-color:yellow; color:black">CHANGED HERE</span>

In [ ]:
# changed all \t to |
def divide_into_verses(text):
    # WEIRD ENCODING: replace weird ‘ with ' accounts for letters before and after
    text = re.sub(r"(\w)‘(\w)", r"\1'\2", text, flags=re.MULTILINE)
    text = re.sub(r"‘", r"'", text, flags=re.MULTILINE)  # for standalone
    # WEIRD ENCODING: replace ’ with '
    text = re.sub(r"(\w)’(\w)", r"\1'\2", text, flags=re.MULTILINE)
    text = re.sub(r"’", r"'", text, flags=re.MULTILINE)  # for standalone
    # WEIRD ENCODING: replace weird ” and “ with " and accounts for letters before and after
    text = re.sub(r"(\w)”", r'\1"', text, flags=re.MULTILINE)
    text = re.sub(r"“(\w)", r'"\1', text, flags=re.MULTILINE)
    text = re.sub(r"”", r'"', text, flags=re.MULTILINE)  # for standalone
    text = re.sub(r"“", r'"', text, flags=re.MULTILINE)  # for standalone

    # SPACING: remove double spaces
    text = re.sub(r' {2,}', ' ', text, flags=re.MULTILINE)

    # SPACING: remove leading newlines
    text = re.sub(r'^\s*$', r'', text, flags = re.MULTILINE)

    # SPACING: remove leading whitespace (e.g., " Text" -> "Text")
    text = re.sub(r'^\s+', r'', text, flags=re.MULTILINE)

    #BASIC SPLITTING
    text = re.sub(r'^\d+ (\d+)', r'\1', text, flags = re.MULTILINE)
    text = re.sub(r'(\.\s)(\d+\s)', r'\1\n\2', text, flags=re.MULTILINE)
    text = re.sub(r'(\.\s)(\d+)', r'\1\n\2', text, flags=re.MULTILINE)
    
    text = re.sub(r'(\s)(\d+)([A-z])', r'\1\n\2\3', text, flags = re.MULTILINE)
    #for verses next to quotes
    text = re.sub(r'(\s)(\d+\")', r'\1\n\2', text, flags = re.MULTILINE)
    #next to single quotes to the right
    text = re.sub(r'\s(\d+\s\')', r'\n\1', text, flags = re.MULTILINE)
    #double quote before the new verse
    text = re.sub(r'(\")\s(\d+\s)', r'\1\n\2', text, flags = re.MULTILINE)
    
    #colon before new verse
    text = re.sub(r'(:)\s(\d+\s)', r'\1\n\2', text, flags = re.MULTILINE)
    #comma before new verse
    text = re.sub(r'(,)\s(\d+)', r'\1\n\2', text, flags = re.MULTILINE)
    
    #adding tabs for splitting
    text = re.sub(r'(\d+-\d+)', r'\n\1', text, flags=re.MULTILINE)
    text = re.sub(r'(\d+[A-z]-\d+)', r'\1', text, flags = re.MULTILINE)
    text = re.sub(r'(\d+,\d+)', r'\1|', text, flags = re.MULTILINE)

    #split verse ranges from text first
    text = re.sub(r'^(\d+-\d+)', r'\1|', text, flags = re.MULTILINE)
    text = re.sub(r'^(\d+[A-z]-\d+)', r'\1|', text, flags = re.MULTILINE)
    
    #splitting the rest of it with tabs
    text = re.sub(r'^(\d+)([A-z][^-])', r'\1|\2', text, flags = re.MULTILINE)
    text = re.sub(r'^(\d+)\s', r'\1|', text, flags = re.MULTILINE)
    text = re.sub(r'^(\d+)\s(\")', r'\1|\2', text, flags = re.MULTILINE)
    text = re.sub(r'^(\d+)([\"\'])', r'\1|\2', text, flags = re.MULTILINE)
    
    text = re.sub(r'^\n', r'', text, flags=re.MULTILINE)
    
    # ----- SPECIFIC QUIRKS BELOW -----
    text = re.sub(r'¶\s', r'|', text, flags = re.MULTILINE)

    # handling upside down ? and !
    text = re.sub(r'(\d+)(\()', r'\1|\2', text, flags=re.MULTILINE)
    text = re.sub(r'(\d+)(\¿)', r'\1|\2', text, flags = re.MULTILINE)
    text = re.sub(r'(\d+)(\¡)', r'\1|\2', text, flags = re.MULTILINE)
    text = re.sub(r'(\d+)(\?)', r'\1|\2', text, flags = re.MULTILINE)
    # handling 12,13 in tuwali-ifugao

    # formatting for csv/xlsx, delimiter is | or tab
    finalText = []
    bookCh = ""
    
    for line in text.splitlines():
        if line:
            if line[0] not in ["0","1","2","3","4","5","6","7","8","9"]:
                bookCh = line.strip()
            else:
                line = bookCh + "|" +line
        finalText.append(line)

    text= '\n'.join(finalText)

    text = re.sub(r'^([SMJLY].*)\d+(\n)', r'', text, flags = re.MULTILINE)
    #text = re.sub(r'^((\w+) (\d+))', r"\2|\3", text, flags=re.MULTILINE)
    text = re.sub(r'^([SMJLY][^|]*?)\s+(\d+)', r'\1|\2', text, flags=re.MULTILINE)
    
    text = "Book|Chapter #|Verse #|Verse\n" + text
    return text

In [38]:
# from test_v2.ipynb
def divide_into_sentences(text):
    # remove any numbers with spaces at the start of lines
    # example: "1 This is a verse -> "This is a verse"
    text = re.sub(r'^\d+\s*', r'', text, flags=re.MULTILINE)
    
    # removes book and chapter headings (e.g., "Matthew 1")
    # ^[A-Za-zÀ-ÖØ-öø-ÿ\s]+ matches the book name (also covers accented letters for other languaes that do have them)
    # \s+ matches the space between the book name and chapter number
    # \d+ matches the chapter number
    # \s*\n? matches any trailing spaces and newline
    text = re.sub(r'^[0-9A-Za-zÀ-ÖØ-öø-ÿ\s]+\d+\s*\n?', '', text, flags=re.MULTILINE)

    # removes verse numbers
    # (?![\d,\.]) makes sure we don't remove numbers that are part of decimals or commas
    text = re.sub(r'\b\d+(?![\d,\.])', '', text)

    # replaces multiple newlines with a single newline
    text = re.sub(r'\n+', r'\n', text)

    # replaces multiple spaces with a single space
    text = re.sub(r'\s+', ' ', text)

    # ensures there is a space after sentence-ending punctuation if followed by a non-space character
    # [.!?] matches sentence ending punctuation
    # ["”’\'\)\]\}] matches any closing quotes, parentheses, or brackets that may follow the punctuation
    # (?=\S) ensures the punctuation is followed by a non-space character
    text = re.sub(r'([.!?]["”’\'\)\]\}]?)(?=\S)', r'\1 ', text)

    # remove spaces after opening punctuation
    text = re.sub(r'([“‘\(\[])\s+', r'\1', text)

    # remove spaces before closing punctuation
    text = re.sub(r'\s+([”’\)\]])', r'\1', text)

    # ensures each sentence starts on a new line
    # [.!?] matches sentence ending punctuation
    # ["\'”’\)\]\}] matches any closing quotes, parentheses, or brackets that may follow the punctuation
    # \s+ matches the whitespace following the punctuation
    text = re.sub(r'([.!?](?:["\'”’\)\]\}]+)?)\s+', r'\1\n', text)

    # This regex keeps letters (including accented), numbers, and common punctuation marks 
    # while removing inconsistent special characters.
    # Basically, if its not in the list of characters, it gets omitted.
    # I had to consult AI with this one in giving me the characters to be included.
    text = re.sub(r'[^A-Za-zÀ-ÖØ-öø-ÿ\u00C0-\u024F\u1E00-\u1EFF0-9\s\.\,\!\?\:\;\'"“”‘’\-\(\)]', "", text)
    
    # removes leading whitespace (e.g., " Text" -> "Text")
    text = re.sub(r'^\s+', r'', text, flags=re.MULTILINE)

    # brings closing parentheses to the previous line if they are on a new line
    text = re.sub(r'\n\)', r')\n', text)

    # removes white space at the start of each line after ", ', or )"
    text = re.sub(r'^([\"\'\)])\s+', r'\1', text, flags=re.MULTILINE)

    # removes leading whitespace (e.g., " Text" -> "Text")
    text = re.sub(r'^\s+', r'', text, flags=re.MULTILINE)

    # standardizes quotes to " and ' at the end
    text = re.sub(r'[“”]', '"', text)
    text = re.sub(r'[‘’]', "'", text)

    return text

NOTE: Please delete the cleaned folder before clicking `Run All`. 

In [39]:
# data_folder = Path("data") already exists above
cleaned_folder = Path("data/cleaned")
cleaned_folder.mkdir(parents=True, exist_ok=True) 

We will run `divide_into_verses` on the raw webscraping .txts in `/data`. This will be used for the Excel sheets later in the project.

In [40]:
for lang in languages:
    data_folder = Path("data")
    file_path = data_folder / f"{lang}.txt"
    output_path = cleaned_folder / f"{lang}-cleaned.txt"

    # SKIP if the input file does not exist or is empty
    if not file_path.exists() or file_path.stat().st_size == 0:
        print(f"Skipped: {file_path} (file does not exist or is empty)")
        continue

    # read input
    with file_path.open("r", errors="ignore", encoding="utf-8") as f:
        text = f.read()

    # clean the text
    cleaned_text = divide_into_verses(text)

    # save
    with output_path.open("w", encoding="utf-8") as f:
        f.write(cleaned_text)

    print(f"Cleaned and saved: {output_path}")

Cleaned and saved: data\cleaned\spanish-cleaned.txt
Cleaned and saved: data\cleaned\tagalog-cleaned.txt
Cleaned and saved: data\cleaned\english-cleaned.txt
Cleaned and saved: data\cleaned\hiligaynon-cleaned.txt
Cleaned and saved: data\cleaned\bikol-cleaned.txt
Cleaned and saved: data\cleaned\waray-cleaned.txt
Cleaned and saved: data\cleaned\ilocano-cleaned.txt
Cleaned and saved: data\cleaned\cebuano-cleaned.txt
Cleaned and saved: data\cleaned\kapampangan-cleaned.txt
Cleaned and saved: data\cleaned\pangasinense-cleaned.txt
Cleaned and saved: data\cleaned\yakan-cleaned.txt
Cleaned and saved: data\cleaned\ivatan-cleaned.txt
Cleaned and saved: data\cleaned\tausug-cleaned.txt
Cleaned and saved: data\cleaned\yami-cleaned.txt
Cleaned and saved: data\cleaned\tuwali_ifugao-cleaned.txt
Cleaned and saved: data\cleaned\masbateno-cleaned.txt


Next, we will run `divide_into_sentences` on the raw data.

In [41]:
for lang in languages:
    data_folder = Path("data")
    file_path = data_folder / f"{lang}.txt"
    output_path = cleaned_folder / f"(sentences)-{lang}-cleaned.txt"

    # SKIP if the input file does not exist or is empty
    if not file_path.exists() or file_path.stat().st_size == 0:
        print(f"Skipped: {file_path} (file does not exist or is empty)")
        continue

    # read input
    with file_path.open("r", errors="ignore", encoding="utf-8") as f:
        text = f.read()

    # clean the text
    cleaned_text = divide_into_sentences(text)

    # save
    with output_path.open("w", encoding="utf-8") as f:
        f.write(cleaned_text)

    print(f"Cleaned and saved: {output_path}")

Cleaned and saved: data\cleaned\(sentences)-spanish-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-tagalog-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-english-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-hiligaynon-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-bikol-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-waray-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-ilocano-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-cebuano-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-kapampangan-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-pangasinense-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-yakan-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-ivatan-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-tausug-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-yami-cleaned.txt
Cleaned and saved: data\cleaned\(sentences)-tuwali_ifugao-cleaned.txt
Cleaned and saved: data\cleaned\(

Here, we will be getting the word count of the raw `lang.txt` files, freshly scraped from Bible.com, and then a word count of the `(sentences)-lang.txt` files.

In [42]:
data_folder = Path("data")
total_word_count = 0

print("PRE-CLEAN WORD COUNT")

for lang in languages:
    cleaned_file_path = data_folder / f"{lang}.txt"  

    if cleaned_file_path.exists() and cleaned_file_path.stat().st_size > 0:
        with cleaned_file_path.open("r", encoding="utf-8") as f:
            text = f.read()
            word_count = len(text.split())
            total_word_count += word_count
            print(f"Word count for {lang}.txt: {word_count}")
    else:
        print(f"Skipped: {cleaned_file_path} (file does not exist or is empty)")

print(f"Total word count across all -cleaned files: {total_word_count}")

PRE-CLEAN WORD COUNT
Word count for spanish.txt: 76085
Word count for tagalog.txt: 86888
Word count for english.txt: 81210
Word count for hiligaynon.txt: 92970
Word count for bikol.txt: 79624
Word count for waray.txt: 89081
Word count for ilocano.txt: 73149
Word count for cebuano.txt: 84973
Word count for kapampangan.txt: 84635
Word count for pangasinense.txt: 77748
Word count for yakan.txt: 84123
Word count for ivatan.txt: 97534
Word count for tausug.txt: 113208
Word count for yami.txt: 107967
Word count for tuwali_ifugao.txt: 82580
Word count for masbateno.txt: 88302
Total word count across all -cleaned files: 1400077


In [43]:
data_folder = Path("data/cleaned")
total_word_count = 0

print("SENTENCE WORD COUNT")

for lang in languages:
    sentence_file_path = data_folder / f"(sentences)-{lang}-cleaned.txt"  

    if sentence_file_path.exists() and sentence_file_path.stat().st_size > 0:
        with sentence_file_path.open("r", encoding="utf-8") as f:
            text = f.read()
            word_count = len(text.split())
            total_word_count += word_count
            print(f"Word count for {lang}-cleaned.txt: {word_count}")
    else:
        print(f"Skipped: {sentence_file_path} (file does not exist or is empty)")

print(f"Total word count across all (sentences)-cleaned files: {total_word_count}")

SENTENCE WORD COUNT
Word count for spanish-cleaned.txt: 75318
Word count for tagalog-cleaned.txt: 86596
Word count for english-cleaned.txt: 80426
Word count for hiligaynon-cleaned.txt: 92876
Word count for bikol-cleaned.txt: 79543
Word count for waray-cleaned.txt: 88941
Word count for ilocano-cleaned.txt: 73030
Word count for cebuano-cleaned.txt: 84773
Word count for kapampangan-cleaned.txt: 84254
Word count for pangasinense-cleaned.txt: 77670
Word count for yakan-cleaned.txt: 83832
Word count for ivatan-cleaned.txt: 97250
Word count for tausug-cleaned.txt: 113228
Word count for yami-cleaned.txt: 107672
Word count for tuwali_ifugao-cleaned.txt: 82321
Word count for masbateno-cleaned.txt: 88228
Total word count across all (sentences)-cleaned files: 1395958


We can get the word count for the sentences here.

SUMMARY:
| Target Language      | Link to Source      | # Words Before Cleaning | # Words After Cleaning |
| ------------- | ------------- | ------------- |  ------------- |
| Spanish  | https://www.bible.com/versions/1076-jbs-biblia-del-jubileo | 76085 | 75318 |
| Tagalog | https://www.bible.com/versions/177-tlab-ang-biblia | 86888 | 86596 |
| English | https://www.bible.com/versions/3523-nrsvue-new-revised-standard-version-updated-edition-2021 |  81210 | 80426 |
| Hiligaynon/Ilonggo | https://www.bible.com/versions/2190-mbbhil12-maayong-balita-nga-biblia-2012 | 92970 | 92876 |
| Bikol/Bikolano | https://www.bible.com/versions/890-mbbbik92-marahay-na-bareta-biblia | 79624 | 79543 |
| Waray | https://www.bible.com/versions/2198-mbbsam-samarenyo-meaning-based-bible-1984 | 89081 | 88941 |
| Ilocano | https://www.bible.com/versions/782-ripv-ti-baro-a-naimbag-a-damag-biblia | 73149 | 73030 |
| Cebuano | https://www.bible.com/versions/562-rcpv-ang-bag-ong-maayong-balita-biblia | 84973 | 84773 |
| Kapampangan | https://www.bible.com/versions/1141-pmpv-ing-mayap-a-balita-biblia | 84635 | 84254 |
| Pangasinense | https://www.bible.com/versions/2194-mbbpan83-maung-a-balita-biblia | 77748 | 77670 |
| Yakan | https://www.bible.com/versions/1388-yakv-yakan | 84123 | 83832 |
| Ivatan | https://www.bible.com/versions/1315-vtsp-ivatan | 97534 | 97250 |
| Tausug  | https://www.bible.com/versions/1319-tsg-kitab-injil | 113208 | 113228 |
| Yami  | https://www.bible.com/versions/2364-snt-seysyo-no-tao | 107967 | 107672 |
| Tuwali Ifugao | https://www.bible.com/versions/2123-ifkwb-nan-kalin-apu-dios | 82580 | 82321 |
| Masbateño/Masbatenyo | https://www.bible.com/versions/1222-msb-masbatenyo | 88302 | 88228 |
| TOTAL | 16 | 1400077 | 1395958 |

Using `pandas`, we can turn the verse txt file into an excel/xlsx file.

[Source](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.ExcelWriter.html)

In [44]:
# data_folder = Path("data/cleaned")
output_excel_path = Path("verses.xlsx")

# if verses.xlsx exists just delete it
if output_excel_path.exists():
    print("unlinking/deleting old version of verses.xlsx")
    output_excel_path.unlink()

with pd.ExcelWriter(output_excel_path) as pd_writer:
    for lang in languages:
        print(f"attempting {lang}")
        cleaned_file_path = data_folder / f"{lang}-cleaned.txt"

        if cleaned_file_path.exists() and cleaned_file_path.stat().st_size > 0:
            try: 
                with cleaned_file_path.open("r", encoding="utf-8") as f:
                    df = pd.read_csv(cleaned_file_path, sep='|', on_bad_lines='error') 
                    excel = df.to_excel(pd_writer, sheet_name=f'{lang}-verses', index=False) # false indexing bc we already have it
                    print(f"Added {lang} to verses.xlsx")
            except Exception as e:
                print(f"Error processing {lang}: {e}")
    print("Finished")

unlinking/deleting old version of verses.xlsx
attempting spanish
Error processing spanish: Error tokenizing data. C error: Expected 4 fields in line 92, saw 5

attempting tagalog


Error processing tagalog: Error tokenizing data. C error: Expected 4 fields in line 1804, saw 5

attempting english
Error processing english: Error tokenizing data. C error: Expected 4 fields in line 1571, saw 5

attempting hiligaynon
Error processing hiligaynon: Error tokenizing data. C error: Expected 4 fields in line 371, saw 5

attempting bikol
Error processing bikol: Error tokenizing data. C error: Expected 4 fields in line 1007, saw 5

attempting waray
Error processing waray: Error tokenizing data. C error: Expected 4 fields in line 1473, saw 5

attempting ilocano
Error processing ilocano: Error tokenizing data. C error: Expected 4 fields in line 1825, saw 5

attempting cebuano
Error processing cebuano: Error tokenizing data. C error: Expected 4 fields in line 1130, saw 5

attempting kapampangan
Error processing kapampangan: Error tokenizing data. C error: Expected 4 fields in line 2119, saw 5

attempting pangasinense
Error processing pangasinense: Error tokenizing data. C error:

## 3. Parallel Corpus
*Create a parallel corpus organized by verse. The number of language pairs should be equal to 3 + n, where n is the number of group member. Example: TGL-CEB, CEB-ILO, ILO-WAR, TGL-WAR.*

## X. AI Declaration


## X. Resources